src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
 
         // Priority is the ExpressionSyntax. It might return ExpressionSyntax which might be a constant pattern such as 'case 3:' 
         // All constant expressions are converted to the constant pattern in the switch binder if it is a match statement.
-        // It is used for parsing patterns in the switch cases. It never returns constant pattern!
-        private CSharpSyntaxNode ParseExpressionOrPatternForCase()
+        // It is used for parsing patterns in the switch cases. It never returns constant pattern, because for a `case` we
+        // need to use a pre-pattern-matching syntax node for a constant case.
+        private CSharpSyntaxNode ParseExpressionOrPattern(bool forCase)
         {
-            var tk = this.CurrentToken.Kind;
-            CSharpSyntaxNode node = null;
+            // handle common error recovery situations during typing
+            switch (this.CurrentToken.Kind)
+            {
+                case SyntaxKind.CommaToken:
+                case SyntaxKind.SemicolonToken:
+                case SyntaxKind.CloseBraceToken:
+                case SyntaxKind.CloseParenToken:
+                case SyntaxKind.CloseBracketToken:
+                case SyntaxKind.EqualsGreaterThanToken:
+                    return this.ParseIdentifierName(ErrorCode.ERR_MissingArgument);
```

Comment:
>ErrorCode.ERR_MissingArgument [](start = 52, length = 29)

It is not clear why this error would always be appropriate. After the ```case```, for example. Do we have tests for scenarios like that? Even in other cases, do we expect at least a pattern?

Reply:
This is only called where a pattern would be appropriate. Probably a new diagnostic about a missing pattern would be better. Recorded in [https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222) .

---
In reply to: [151811067](https://github.com/dotnet/roslyn/pull/23208#discussion_r151811067) [](ancestors = 151811067)